### ISSUE TITLE
QUESTION: how does Hyperopt optimize

### ISSUE THREAD

**Issue by Vildnex** [freqtrade/freqtrade#3384]
Hey guys, I'm trying to understand Hyperopt and how it works, and I'm getting a little bit confused.

1. In terms of loss function, if I will use **SharpeHyperOptLossDaily** or **SortinoHyperOptLossDaily** that means if going to optimize my strategy on daily results, right?
2. If I will use `--min-trades=10` that means in combination with the loss function from above, it's going to get at least 10 trades per day? Or this is available for the whole period?
3. What's considered in terms of trained data and epochs overfitting? If I will train mine with 10.000 epochs with around 700days of data(5min timeframe) that's ok or it's too much? 
4. How will be the actually process of training?

- **[NO forward optimization]** it's going to use all the data once (let's say 300 days) and try to minimize the loss function based on those?
- **[YES forward optimization]** Or it's going to split the data into some parts (let's say 100day each chunk) and it's going to train the first 100, then the second 100 and so on and so forth.

**Comment by hroff-1902** [comment]
1. Yes, it will use Sharpe and Sortino metrics as the values of the objective function, calculated on returns aggregated on daily basis.
2. No, with `--min-trades 10` hyperopt substitutes the actual value of the objective function for epochs with number of trades less than 10 with some large value, marking such epochs as "bad", increasing the probability to avoid the neighbor points in the hyperspace from the optimization path.
3. I do not think that the term "training" is applicable to the process of optimization at all. Hyperopt does not work on different sets of data, it works on single set of data for the chosen pairlist/timerange and searches for a local minimum of the objective (loss) function on this set of data, varying the values of hyperspace dimensions.
As for the number of epochs, hyperopt uses bayesian search, which quickly descends to one of the local minimums of the objective function surface in the hyperspace, it's not grid search, you don't need to run hyperopt for 10000 epochs -- you simply won't find better results after, say, several hundreds of epochs. I personally never run hyperopt for more than 500-700-1000 epochs. So instead of running it for 10000 epochs you can better run it 20 times for 500 epochs or 10 times for 1000 epochs -- it will give you 20 or 10 different local minimums of the object function respectively, which you can then select from using some external criteria..
4. I think I did not get it. If you need to verify hyperopt results on other timerange -- transfer results from hyperopt to the strategy and backtest it on the different set of data.

**Comment by xmatthias** [comment]
for 3 and 4:
You might want to look at the `--timerange`  parameter. It defines how much (or little) data is passed to hyperopt - so you might want to reserve a certain percentage of days for a "validation" run once you got your found parameters transferred to backtesting. (obviously confirm backtesting results  on the same timerange to ensure you've transferred the results correctly).

Hyperopt does not do any feed/forward / splitting of data by itself - it'll feed whatever data is there (or limited by `--timerange`).

Hope this helps answer the remaining points.

**Comment by Vildnex** [comment]
That makes a lot of sense, thanks for answering me @xmatthias  and @hroff-1902  :D
